Enhance pipeline TFX taxi sample to support on-prem cluster

jinchihe opened the pull request:
fixes: #721

Enhance pipeline TFX taxi sample to support on-prem cluster

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kubeflow/pipelines/749)
<!-- Reviewable:end -->

jinchihe commented:
/assign @IronPan

IronPan commented:
/assign @gaoning777 @Ark-kun

gyliu513 commented on samples/tfx/README.md:
```diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
 ## Requirements
 
-Preprocessing and model analysis use [Apache Beam](https://beam.apache.org/).
+- Using GCP
```
s/- Using GCP/### Using GCP

jinchihe replied:
@gyliu513  Addressed.

jinchihe replied:
> s/- Using GCP/### Using GCP

Done

gyliu513 commented on samples/tfx/README.md:
```diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
-When run with the `cloud` mode (instead of the `local` mode), those steps use [Google Cloud DataFlow](https://beam.apache.org/) for running the Beam pipelines.
+  Therefore, you must enable the DataFlow API for the given GCP project if you want to use `cloud` as the mode for either preprocessing or analysis. See the [guide to enabling the DataFlow API](https://cloud.google.com/endpoints/docs/openapi/enable-api).
+
+- On-prem cluster
```
s/- On-prem cluster/### On-prem cluster

jinchihe replied:
Done

jinchihe replied:
> s/- On-prem cluster/### On-prem cluster

Done

gyliu513 commented on samples/tfx/README.md:
````diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
-```bash
-dsl-compile --py taxi-cab-classification-pipeline.py --output taxi-cab-classification-pipeline.tar.gz
-```
+- GCP
````
s/- GCP/### GCP

jinchihe replied:
Done

jinchihe replied:
> s/- GCP/### GCP

Done

gyliu513 commented on samples/tfx/README.md:
````diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
+  ```bash
+  dsl-compile --py taxi-cab-classification-pipeline-gcp.py --output taxi-cab-classification-pipeline.tar.gz
+  ```
+- On-prem cluster
````
s/- On-prem cluster/### On-prem cluster

jinchihe replied:
> s/- On-prem cluster/### On-prem cluster

Done

gyliu513 commented on backend/src/apiserver/config/sample_config.json:
```diff
@@ -7,7 +7,7 @@
   {
     "name":"[Sample] ML - TFX - Taxi Tip Prediction Model Trainer",
     "description":"Example pipeline that does classification with model analysis based on a public tax cab BigQuery dataset. For source code, refer to https://github.com/kubeflow/pipelines/tree/master/samples/tfx",
```
How about add both gcp and on-repm tar.gz to the UI?

swiftdiaries replied:
+1

jinchihe replied:
> How about add both gcp and on-repm tar.gz to the UI?

Done

gyliu513 commented on samples/tfx/README.md:
```diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
+
+- On-prem cluster
+
+  When run the on-prem clusters, follow the [document](https://kubernetes.io/docs/concepts/storage/persistent-volumes/) to create the persistent volume and persistent volume claim to storage the intermediate data and result. Note that the `accessModes` should be `ReadWriteMany` so that the volume can be mounted as read-write by many nodes. For example, the `taxi-cab-classification-pipeline_on-pre ...
```
s/taxi-cab-classification-pipeline_on-prem.py/taxi-cab-classification-pipeline-on-prem.py

jinchihe replied:
> s/taxi-cab-classification-pipeline_on-prem.py/taxi-cab-classification-pipeline-on-prem.py

Done

gyliu513 commented on samples/tfx/README.md:
````diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
-```
+- GCP
+  ```bash
+  dsl-compile --py taxi-cab-classification-pipeline-gcp.py --output taxi-cab-classification-pipeline.tar.gz
````
s/taxi-cab-classification-pipeline.tar.gz/taxi-cab-classification-pipeline-gcp.tar.gz

jinchihe replied:
> s/taxi-cab-classification-pipeline.tar.gz/taxi-cab-classification-pipeline-gcp.tar.gz

Done

gyliu513 commented on samples/tfx/README.md:
````diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
+  ```
+- On-prem cluster
+  ```bash
+  dsl-compile --py taxi-cab-classification-pipeline-on-prem.py --output taxi-cab-classification-pipeline.tar.gz
````
s/taxi-cab-classification-pipeline.tar.gz/taxi-cab-classification-pipeline-om-prem.tar.gz

jinchihe replied:
> s/taxi-cab-classification-pipeline.tar.gz/taxi-cab-classification-pipeline-om-prem.tar.gz

Done

gyliu513 commented on test/sample-test/run_test.sh:
```diff
@@ -169,17 +169,17 @@ elif [ "$TEST_NAME" == "tfx" ]; then
   fi
 
-  dsl-compile --py taxi-cab-classification-pipeline.py --output taxi-cab-classification-pipeline.tar.gz
+  dsl-compile --py taxi-cab-classification-pipeline-gcp.py --output taxi-cab-classification-pipeline.tar.gz
```
s/taxi-cab-classification-pipeline.tar.gz/taxi-cab-classification-pipeline-gcp.tar.gz

jinchihe replied:
> s/taxi-cab-classification-pipeline.tar.gz/taxi-cab-classification-pipeline-gcp.tar.gz

Done

gyliu513 commented on test/sample-test/run_test.sh:
```diff
@@ -169,17 +169,17 @@ elif [ "$TEST_NAME" == "tfx" ]; then
-    sed -i -e "s|gcr.io/ml-pipeline/ml-pipeline-kubeflow-deployer:\([a-zA-Z0-9_.-]\)\+|${KUBEFLOW_DEPLOYER_IMAGE}|g" taxi-cab-classification-pipeline.py
-    sed -i -e "s|gcr.io/ml-pipeline/ml-pipeline-local-confusion-matrix:\([a-zA-Z0-9_.-]\)\+|${LOCAL_CONFUSIONMATRIX_IMAGE}|g" taxi-cab-classification-pipeline.py
-    sed -i -e "s|gcr.io/ml-pipeline/ml-pipeline-local-roc:\([a-zA-Z0-9_.-]\)\+|${LOCAL_ROC_IMAGE}|g" taxi-cab-classification-pipeline.py
+    sed -i -e "s|gcr.io/ml-pipeline/ml-pipeline-dataflow-tft:\([a-zA-Z0-9_.-]\)\+|${DATAFLOW_TFT_IMAGE}|g" taxi-cab-classification-pipeline-gcp.py
```
How to trigger the test for `taxi-cab-classification-pipeline-on-prem.py`?

jinchihe replied:
> How to trigger the test for `taxi-cab-classification-pipeline-on-prem.py`?

@gyliu513 I added TODO comments in the run_test.sh. We may create testing case in the future. Thanks.

jinchihe commented:
/retest

jinchihe commented:
/retest

jinchihe commented:
/retest

jinchihe commented:
/test kubeflow-pipeline-e2e-test

neuromage commented:
Thanks for the PR!

Overall, the direction to support on-prem is great. However, I'm not in favour of having two versions of the pipeline, one for GCP and one using NFS, as it increases maintenance cost. Ideally, we should have one pipeline, with a user-configurable option of whether to store artifacts on GCP or NFS. Would it be feasible to explore this approach instead?

gaoning777 commented:
echoed @neuromage

gyliu513 commented:
@neuromage good point, let me check with @jinchihe for how we can move forward.

jinchihe commented:
> Overall, the direction to support on-prem is great. However, I'm not in favour of having two versions of the pipeline, one for GCP and one using NFS, as it increases maintenance cost. Ideally, we should have one pipeline, with a user-configurable option of whether to store artifacts on GCP or NFS. Would it be feasible to explore this approach instead?

@neuromage @gaoning777 @gyliu513 

In fact, we discussed the solution in the original issue #721, as @swiftdiaries mentioned in #721, it's better to have separate example code. The onprem wouldn't be using gcp secrets so having another tfx_taxi_cab_onprem.py makes sense. And we'd be needing volumes, so the code would look very different.

vicaire commented:
@jinchihe, what do you think of the DSL changes propose at https://github.com/kubeflow/pipelines/issues/801 to support volumes in the DSL?

jinchihe commented:
@vicaire Great for supporting local storage easily. So I think we can hold on the PR then. Thanks.

vicaire commented:
OK thanks. Let's conclude the discussion on #801 and hold on this PR till then.

cvenets commented:
Hello @jinchihe, I was testing your PR today, trying to run the whole example completely on-prem.

Although everything looks good, I'm getting an error in the last step of the pipeline, the `kubeflow_deploy_op`. The log states:

```
+ KUBERNETES_NAMESPACE=kubeflow
+ SERVER_NAME=model-server
+ (( 8 ))
+ case $1 in
+ shift
+ CLUSTER_NAME=tfx-taxi-pipeline-on-prem
+ shift
+ (( 6 ))
+ case $1 in
+ shift
+ MODEL_PATH=/mnt/tfx-taxi-cab-classification-pipeline-example-z85z5/train
+ shift
+ (( 4 ))
+ case $1 in
+ shift
+ SERVER_NAME=taxi-cab-classification-model-tfx-taxi-cab-classification-pipeline-example-z85z5
+ shift
+ (( 2 ))
+ case $1 in
+ echo 'Unknown argument: '\''--pvc-name'\'''
Unknown argument: '--pvc-name'
+ exit 1
```

After digging in a bit, looking into the container image, I understand that the container image for the deploy part, which in your code is `image='gcr.io/ml-pipeline/ml-pipeline-kubeflow-deployer:be19cbc2591a48d2ef5ca715c34ecae8223cf454'`, contains a `deploy.sh` script that doesn't support either the `--model-storage-type` or `--pvc-name` arguments. Thus the error.

Can you point me to the image you used to run the sample successfully?

Thank you,
Constantinos

jinchihe commented:
> Hello @jinchihe, I was testing your PR today, trying to run the whole example completely on-prem.
> 
> Although everything looks good, I'm getting an error in the last step of the pipeline, the `kubeflow_deploy_op`. The log states:
> 
> ```
> + KUBERNETES_NAMESPACE=kubeflow
> + SERVER_NAME=model-server
> + (( 8 ))
> + case $1 in
> + shift
> + CLUSTER_NAME=tfx-taxi-pipeline-on-prem
> + shift
> + (( 6 ))
> + case $1 in
> + shift
> + MODEL_PATH=/mnt/tfx-taxi-cab-classification-pipeline-example-z85z5/train
> + shift
> + (( 4 ))
> + case $1 in
> + shift
> + SERVER_NAME=taxi-cab-classification-model-tfx-taxi-cab-classification-pipeline-example-z85z5
> + shift
> + (( 2 ))
> + case $1 in
> + echo 'Unknown argument: '\''--pvc-name'\'''
> Unknown argument: '--pvc-name'
> + exit 1
> ```
> 
> After digging in a bit, looking into the container image, I understand that the container image for the deploy part, which in your code is `image='gcr.io/ml-pipeline/ml-pipeline-kubeflow-deployer:be19cbc2591a48d2ef5ca715c34ecae8223cf454'`, contains a `deploy.sh` script that doesn't support either the `--model-storage-type` or `--pvc-name` arguments. Thus the error.
> 
> Can you point me to the image you used to run the sample successfully?
> 
> Thank you,
> Constantinos

Thanks for the tests @cvenets. The problem has been fixed in #755, due to the RP is proposed to hold on, the image version is not updated. 
I just rebased and updated to new image. That's should be OK now. :-)

gyliu513 commented on test/sample-test/run_test.sh:
```diff
@@ -168,25 +168,27 @@ elif [ "$TEST_NAME" == "tfx" ]; then
   SAMPLE_SEQUENTIAL_TEST_RESULT=junit_SampleSequentialOutput.xml
   SAMPLE_SEQUENTIAL_TEST_OUTPUT=${RESULTS_GCS_DIR}
 
+  #TODO: Create testing for sample TFX taxi on-prem cluster.
```
s/#TODO:/# TODO:

gyliu513 commented on samples/tfx/taxi-cab-classification-pipeline-on-prem.py:
```diff
@@ -0,0 +1,200 @@
+
+
+@dsl.pipeline(
+    name='TFX Taxi Cab Classification Pipeline Example',
```
Shall we add `on-prem` for name and description here? Also does the pipeline gcp file also need some update for name and description?

gyliu513 commented on samples/tfx/README.md:
```diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
 ## Requirements
 
-Preprocessing and model analysis use [Apache Beam](https://beam.apache.org/).
+### Using GCP
```
s/### Using GCP/### GCP

gyliu513 commented on samples/tfx/README.md:
```diff
@@ -19,30 +19,54 @@ dataset in [Google BigQuery](https://cloud.google.com/bigquery/). Explore the
-Preprocessing and model analysis use [Apache Beam](https://beam.apache.org/).
+### Using GCP
+
+  Preprocessing and model analysis use [Apache Beam](https://beam.apache.org/).
```
Do not need blank space at the beginning, ditto for others.

gyliu513 commented on samples/tfx/README.md:
```diff
@@ -61,3 +85,4 @@ Analysis:
 Prediction:
   [source code](https://github.com/kubeflow/pipelines/tree/master/components/dataflow/predict/src) 
   [container](https://github.com/kubeflow/pipelines/tree/master/components/dataflow/predict)
+
```
kill this

cvenets commented:
Hello @jinchihe, thank you for updating the pipeline code and image.

Unfortunately, when trying to run the updated pipeline the first step (validation) now breaks with the following log. This didn't happen with the previous pipeline code, all steps finished successfully except for the deployment part. Any thoughts?

I see that I'm running the updated image now: `image: gcr.io/ml-pipeline/ml-pipeline-dataflow-tfdv:5df2cdc1ed145320204e8bc73b59cdbd7b3da28f`

```
tfx-taxi-cab-classification-pipeline-example-for-on-prem-cqg2mn-3535214929
This step is in Failed state with this message: failed with exit code 1
    from tensorflow_transform.coders.csv_coder import CsvCoder
  File "/usr/local/lib/python2.7/dist-packages/tensorflow_transform/coders/csv_coder.py", line 26, in <module>
    import tensorflow as tf
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/__init__.py", line 22, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/__init__.py", line 81, in <module>
    from tensorflow.python import keras
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/__init__.py", line 24, in <module>
    from tensorflow.python.keras import activations
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/activations/__init__.py", line 22, in <module>
    from tensorflow.python.keras._impl.keras.activations import elu
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/_impl/keras/__init__.py", line 21, in <module>
    from tensorflow.python.keras._impl.keras import activations
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/_impl/keras/activations.py", line 23, in <module>
    from tensorflow.python.keras._impl.keras import backend as K
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/_impl/keras/backend.py", line 37, in <module>
    from tensorflow.python.layers import base as tf_base_layers
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/layers/base.py", line 25, in <module>
    from tensorflow.python.keras.engine import base_layer
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/engine/__init__.py", line 21, in <module>
    from tensorflow.python.keras.engine.base_layer import InputSpec
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/engine/base_layer.py", line 33, in <module>
    from tensorflow.python.keras import backend
  File "/usr/local/lib/python2.7/dist-packages/tensorflow/python/keras/backend/__init__.py", line 22, in <module>
    from tensorflow.python.keras._impl.keras.backend import abs
ImportError: cannot import name abs
```

Thank you,
Constantinos

hongye-sun commented:
Constantinos, the image is broken because of a pip dependency updates.  Could you use an older image for now like: 6ad2601ec7d04e842c212c50d5c78e548e12ddea. I have a PR(#817) to revert the sample code to use the latest image. We have another PR(#828) to fix the broken image issue.

cvenets commented:
Hello @hongye-sun, thank you for your help.

After looking at #817, which I see is now merged, I changed the image version of tfdv, tft, tfma, and tf-predict on this PR's on-prem pipeline, to the version you are suggesting and it completed successfully.

I see that PR #828 is also merged. Does this mean that you will produce a new version for all images, which will work for both GCP and on-prem pipelines?

@jinchihe, does it make sense to update this PR replacing the tfdv, tft, tfma, and tf-serving image version with the one that @hongye-sun suggested, and shows on #817, so that this PR's pipeline is complete and working?

Thank you both for your help.

jinchihe commented:
@cvenets Seems the #817 has been merged. I updated the image version to new one. Thanks.

gaoning777 commented:
Just a heads-up: when a PR that contains component code changes is merged, it does not mean the corresponding component images are released. These images are built and staged when the PR is merged but are released later. However, when you find the image tags updated in the samples, the images are released by then.
Thanks

cvenets commented:
Hello @jinchihe, I see that you updated the version on all images, not just the tfdv, tfma, tft, and tf-predict images, as done in #817. Will that work?

Thank you,
Constantinos

vicaire commented:
Note: assuming that this PR is still pending https://github.com/kubeflow/pipelines/issues/801, please let us know if you think we can proceed with the review.

Ark-kun commented:
@jinchihe
Can you please fix the conflicts and let's merge this one.

jinchihe commented:
> @jinchihe
> Can you please fix the conflicts and let's merge this one.

OK, will do that later once get chance. Thanks.

jinchihe commented:
Seems need more code updates for support on-prem cluster to fit recently changes. I'm going to do that now. Thanks.

jomach commented:
I'm looking at this, and I cannot find the function  for the volumes on the example for this pr.

from kfp import onprem <-- Where is this defined ?

jinchihe commented:
> I'm looking at this, and I cannot find the function for the volumes on the example for this pr.
> 
> from kfp import onprem <-- Where is this defined ?

Please look at here: https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/onprem.py

jinchihe commented:
> @jinchihe
> Can you please fix the conflicts and let's merge this one.

@Ark-kun Could you please review the ticket? Thanks a lot!

ahmed-menshawy commented:
Hello, Thanks so much for providing this on-prem example. I have tried the example but it's failing in the deploy step with this error:

```
+ echo 'Unknown argument: '\''--pvc-name'\'''
+ exit 1
Unknown argument: '--pvc-name'

```

jomach commented:
> > I'm looking at this, and I cannot find the function for the volumes on the example for this pr.
> > from kfp import onprem <-- Where is this defined ?
> 
> Please look at here: https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/onprem.py

Docs where not up to date. 
I created a PR for this here https://github.com/kubeflow/website/pull/688/files

jinchihe commented:
> Hello, Thanks so much for providing this on-prem example. I have tried the example but it's failing in the deploy step with this error:
> 
> ```
> + echo 'Unknown argument: '\''--pvc-name'\'''
> + exit 1
> Unknown argument: '--pvc-name'
> ```

Thanks for the testing. The test is successful on my enviroment as below.
```bash
Name:                tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs
Namespace:           kubeflow
ServiceAccount:      pipeline-runner
Status:              Succeeded
Created:             Thu May 09 23:23:24 -0700 (26 minutes ago)
Started:             Thu May 09 23:23:18 -0700 (26 minutes ago)
Finished:            Thu May 09 23:29:06 -0700 (20 minutes ago)
Duration:            5 minutes 48 seconds
Parameters:          
  project:           taxi-cab-classification-pipeline-on-prem
  pvc-name:          pipeline-pvc
  pvc-mount-path:    /mnt
  relative-column-names: taxi-cab-classification/column-names.json
  key-columns:       trip_start_timestamp
  relative-train:    taxi-cab-classification/train.csv
  relative-evaluation: taxi-cab-classification/eval.csv
  mode:              local
  relative-preprocess-module: taxi-cab-classification/preprocessing.py
  learning-rate:     0.1
  hidden-layer-size: 1500
  steps:             3000
  analyze-slice-column: trip_start_hour

STEP                                                                PODNAME                                                                     DURATION  MESSAGE
 ✔ tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs                                                                                        
 ├-✔ tfx-data-validation                                            tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-281373311   46s       
 ├-✔ transform-using-tf-on-dataflow                                 tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-1772236091  44s       
 ├-✔ train-fc-dnn-using-tf                                          tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-1198291363  3m        
 ├-✔ kubeflow-serve-tf-model                                        tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-2773474228  45s       
 ├-✔ predict-using-tf-on-dataflow                                   tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-4165187364  16s       
 ├-✔ tfx-analyze-model                                              tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-2292059595  33s       
 ├-✔ confusion-matrix                                               tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-2028229238  5s        
 └-✔ roc-curve                                                      tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-2308126162  6s  
```
The parameter was added in #755 and should be included in the new image.
Could you please check image tag of your failed pod, such as:
```bash
[root@jche1 ~]# kubectl describe  pod tfx-taxi-cab-classification-pipeline-example-for-on-prem-cm8mrs-2773474228 |grep -i image
    Image:         gcr.io/ml-pipeline/ml-pipeline-kubeflow-deployer:727c48c690c081b505c1f0979d11930bf1ef07c0
.....
```

ahmed-menshawy commented:
Thanks @jinchihe I used an updated deployer image and it worked fine.

Ark-kun commented on backend/src/apiserver/config/sample_config.json:
```diff
@@ -5,9 +5,14 @@
+    "file":"/samples/tfx/taxi-cab-classification-pipeline-gcp.py.tar.gz"
+  },
+  {
+    "name":"[Sample] ML - TFX - Taxi Tip Prediction Model Trainer for on-prem cluster",
```
Will this sample run out of the box on GKE deployment?

jinchihe replied:
We would like to have the OOB sample for Kubeflow Pipelines. Do you think we should just put one of them(GCP)? if so, I will remove the onprem sample from OOB, thanks.

Ark-kun replied:
@IronPan @paveldournov @gaoning777 @hongye-sun  WDYT?

jinchihe replied:
@Ark-kun Reconsider this, we just show one of them is enough, and with the last solution, need to compile for onprem cluster, will remove this and only keep OOB sample for GCP. Thanks,

jinchihe commented:
Seems `samples/tfx/taxi-cab-classification-pipeline.py` has conflicts again. 

Now I think it's better to merge the `taxi-cab-classification-pipeline-on-prem.py` and `taxi-cab-classification-pipeline-gcp.py` into one file in the PR, if so, easy maintenance and avoid code duplication. @Ark-kun  @gaoning777  WDYT?

Personally I strongly suggest to merge the PR, since so many people pay attention to the PR and ask questions about running the sample on prem (local storage) by Slack and Email. :-)

Ark-kun commented:
>has conflicts again

Sorry. There was a release yesterday and releases currently change the SHA hashes in the component paths.

>Now I think it's better to merge the taxi-cab-classification-pipeline-on-prem.py and taxi-cab-classification-pipeline-gcp.py into one file in the PR, if so, easy maintenance and avoid code duplication. @Ark-kun @gaoning777 WDYT?

It would be great. 
https://github.com/kubeflow/pipelines/pull/1209 will make it much easier.

jinchihe commented:
@Ark-kun Thanks. here's what I'm thinking. Please review and comment. Thanks a lot!

Due to there are some differences for gcp and on-prem cluster for `parameters` and `ContainerOp` , such as I'd like to use the feature in #801 to create a pvc and download source `taxi-cab-classification` to the pvc automatically, and then attach the pvc to each steps by `pvolumes` of ContainerOp, that means each component will has different parameter between gcp and on-prem, as on-prem will add `pvolumes`:

```
    cop = dsl.ContainerOp(
        ...
        pvolumes={"/mnt": vop.volume}
    )
```

So here's what I'm thinking: add new function in `taxi-cab-classification-pipeline.py`, after the change, one function is for gcp and another function is for on-prem, as following:

```
...

@dsl.pipeline(
  name='TFX Taxi Cab Classification Pipeline Example for GCP',
  description='Example pipeline that does classification with model analysis based on a public BigQuery dataset for GCP.'
)
def taxi_cab_classification_gcp(
    output,
    project,
    column_names='gs://ml-pipeline-playground/tfx/taxi-cab-classification/column-names.json',

...

@dsl.pipeline(
  name='TFX Taxi Cab Classification Pipeline Example for on-prem cluster',
  description='Example pipeline that does classification with model analysis based on a public BigQuery dataset for on-prem cluster.'
)
def taxi_cab_classification_onprem(
    project='taxi-cab-classification-pipeline-on-prem',
    relative_column_names='taxi-cab-classification/column-names.json',

...

if __name__ == '__main__':
    kfp.compiler.Compiler().compile(taxi_cab_classification_gcp, __file__ + '_gcp.zip')
    kfp.compiler.Compiler().compile(taxi_cab_classification_onprem, __file__ + '_onprem.zip')
```

And compile with `--function`, as below

- GCP:

```
dsl_compile --py taxi-cab-classification-pipeline.py --output taxi-cab-classification-pipeline_gcp.tar.gz --function taxi_cab_classification_gcp
```
- on premise cluster.

```
dsl_compile --py taxi-cab-classification-pipeline.py --output taxi-cab-classification-pipeline_onprem.tar.gz --function taxi_cab_classification_onprem
```

elikatsis commented:
Hi @jinchihe,
>[...] such as I'd like to use the feature in #801 to create a pvc and download source taxi-cab-classification to the pvc automatically, and then attach the pvc to each steps by pvolumes of ContainerOp, that means each component will has different parameter between gcp and on-prem, as on-prem will add pvolumes: [...]

Thank you for integrating with the work merged at #926 and I'm glad you find `pvolumes` convenient.
Is there something you have bumped into while using `pvolumes`? Please let me know.

I have not used components extensively, what is the easiest way to use a `pvolume` instance with a component? Looking at code examples, I think you'd find an `.add_pvolumes()` method very helpful, so you could do, for example,

```python
vop = dsl.VolumeOp(...)
validation = dataflow_tf_data_validation_op(...).add_pvolumes({pvc_mount_path: vop.volume})
preprocess = dataflow_tf_transform_op(...).add_pvolumes({pvc_mount_path: validation.pvolume})
# or preprocess = dataflow_tf_transform_op(...).add_pvolumes({pvc_mount_path: vop.volume})
# since in this sample all the dependencies are resolved using parameters
...
```
I'm preparing a WIP PR so you can test.

Ark-kun commented:
> I'd like to use the feature in #801 to create a pvc and download source `taxi-cab-classification` to the pvc automatically, and then attach the pvc to each steps by `pvolumes` of ContainerOp

What about creating the volume, but then using `onprem.mount_pvc` to mount it to the steps?

jinchihe commented:
> What about creating the volume, but then using `onprem.mount_pvc` to mount it to the steps?

@Ark-kun Thanks, This method `onprem.mount_pvc` is currently used. If using `dsl.VolumeOp` to create the pvc and then using `onprem.mount_pvc` to mount that, I'm not sure pvc name can passed to `onprem.mount_pvc`. 

The good way is to use `.add_pvolumes()` method which mentioned by @elikatsis (#1353 is on the way), I'm going to try that.

jinchihe commented:
/assign @Ark-kun 

I just updated the PR as following:
1. For easy maintenance and to avoid code duplication, I merged the pipeline definition files for GCP and On-Prem cluster.
2. For On-Prem cluster, I added two steps for creating PVC and download code automatically.
3. For `sample_config.json`, I removed the changes and only keep one simple that's for GCP.
4. Define the variable `platform` to switch GCP and On-Prem, by default that's for GCP, if use want to try the pipeline on On-Prem cluster, need to update simplely as README description.
5. For the code to mount pvc and GCP use_gcp_secret, we can enhance later once the PR #1209 merged, I marked to TBD.

Any more comments please let me know. Thanks a lot!

jinchihe commented:
@Ark-kun Any more comments. I noticed the #1209 has been merged, but I'd like to keep existing implement since some steps no need to apply pvc, and I used for loop to mount that and I think the code is clear  :-)

Ark-kun commented:
/lgtm
/approve

Sorry for the delays!

jinchihe commented:
/retest

cdyangzhenyu commented:
@jinchihe Hi jinchi, I was testing the TFX demo on prem today. The pipeline workflow worked fine, but the web ui has some error, I can't open the artifacts ui. I find some js error: 'Uncaught (in promise) Error: Unsupported storage path: /mnt/a07b8215-8c1a-11e9-a2ff-525400ed33aa/tfx-taxi-cab-classification-pipeline-example-q7rtq-2449399348/data/roc.csv'. 
How to solve this problem? Thanks.

jinchihe commented:
@cdyangzhenyu I think the Pipeline Web UI does not support get the artifacts from local path, but just from gs:// and s3:// etc... May be support this later. Thanks.
